R/archive.R

```diff
@@ -196,572 +187,755 @@ next_after.Date <- function(x) x + 1L
+            "to one > %3$s, implying at least one version in between."
+          ), epi_archive$versions_end, next_version_tag, fill_versions_end))
         }
-        self$DT <- self$DT[self$DT$version <= max_version, colnames(self$DT), with = FALSE]
-        # (^ this filter operation seems to always copy the DT, even if it
-        # keeps every entry; we don't guarantee this behavior in
-        # documentation, though, so we could change to alias in this case)
-        if (!is.na(self$clobberable_versions_start) && self$clobberable_versions_start > max_version) {
-          self$clobberable_versions_start <- NA
+        nonversion_key_vals_ever_recorded <- unique(epi_archive$DT, by = nonversion_key_cols)
+        # In edge cases, the `unique` result can alias the original
+        # DT; detect and copy if necessary:
+        if (identical(address(epi_archive$DT), address(nonversion_key_vals_ever_recorded))) {
+          nonversion_key_vals_ever_recorded <- copy(nonversion_key_vals_ever_recorded)
         }
-        self$versions_end <- max_version
-        return(invisible(self))
+        next_version_dt <- nonversion_key_vals_ever_recorded[
+          , version := next_version_tag
+        ][
+          # this makes the class of these columns logical (`NA` is a
+          # logical NA; we're relying on the rbind below to convert to
+          # the proper class&typeof)
+          , (nonkey_cols) := NA
+        ]
+        # full result DT:
+        setkeyv(rbind(epi_archive$DT, next_version_dt), key(epi_archive$DT))[]
       },
-      #####
-      #' @description Merges another `epi_archive` with the current one, mutating the
-      #'   current one by reseating its `DT` and several other fields, but avoiding
-      #'   mutation of the old `DT`; returns the current archive
-      #'   [invisibly][base::invisible]. See [`epix_merge`] for a full description
-      #'   of the non-R6-method version, which does not mutate either archive, and
-      #'   does not alias either archive's `DT`.
-      #' @param y as in [`epix_merge`]
-      #' @param sync as in [`epix_merge`]
-      #' @param compactify as in [`epix_merge`]
-      merge = function(y, sync = c("forbid", "na", "locf", "truncate"), compactify = TRUE) {
-        result <- epix_merge(self, y,
-          sync = sync,
-          compactify = compactify
-        )
-
-        if (length(epi_archive$private_fields) != 0L) {
-          cli_abort("expected no private fields in epi_archive",
-            internal = TRUE
-          )
-        }
+      "locf" = {
+        # just the old DT; LOCF is built into other methods:
+        epi_archive$DT
+      }
+    )
+    new_versions_end <- fill_versions_end
+    # Update `epi_archive` all at once with simple, error-free operations +
+    # return below:
+    epi_archive$DT <- new_dt
+    epi_archive$versions_end <- new_versions_end
+  } else {
+    # Already sufficiently up to date; nothing to do.
+  }
+  return(invisible(epi_archive))
```

Comment:
todo: return visibly, not invisibly.  Same deal with the rest of the S3 methods besides `print`.  These other methods are no longer used primarily for mutation or another side effect, but rather are primarily for a return value, so return value should be visible.